src/plugins/hidsim/gui/vmi_win_gui_parser.cpp

```diff
@@ -0,0 +1,1014 @@
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ * This file was created by Jan Gruber.                                    *
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include <wchar.h>
+#include <string.h>
+
+/* Offset retrieval */
+#include "vmi_win_gui_offsets.h"
+#include "vmi_win_gui_atoms.h"
+#include "vmi_win_gui_utils.h"
+#include "vmi_win_gui_filter_definitions.h"
+
+#include "vmi_win_gui_parser.h"
+
+struct rect* get_visible_rect_from_bitmask(
+    char* map, size_t n, int scanline, struct rect* r)
+{
+    struct rect* result = NULL;
+
+    size_t byte;
+    int bit_idx;
+    unsigned int bit;
+    int x0 = -1, x1 = -1, y0 = -1, y1 = -1;
+    bool is_y_hole = false;
+    bool is_x_hole = false;
+
+    for (int y = r->y0; y < r->y1; y++)
+    {
+        for (int x = r->x0; x < r->x1; x++)
+        {
+
+            byte = (y * scanline + x)/8;
+
+            /* Parts of a wnd can be outside of the desktop's frame */
+            if (byte >= n || x < 0)
+                continue;
+
+            bit_idx = x % 8;
+            bit = 0x80 >> bit_idx;
+            if (!(map[byte] & bit))
+            {
+                if (x0 == -1 && y0 == -1)
+                {
+                    x0 = x;
+                    y0 = y;
+                }
+
+                if (!is_x_hole)
+                    x1 = x;
+
+                if (!is_y_hole)
+                    y1 = y;
+            }
+            else
+            {
+                if (x == x0)
+                    is_y_hole = true;
+
+                if (y == y1)
+                    is_x_hole = true;
+            }
+        }
+    }
+    if (x1 != -1 && y1 != -1)
+    {
+        result = (struct rect*) malloc(sizeof(struct rect));
+        result->x0 = x0;
```

Comment:
malloc result check?

Reply:
Added additional checks for allocation failures, which sneaked through my search and replace in bb3b7867851265325991cdc23361e049ab5cf85f and bc66be92ca943a6d4a6472e76c10c0801b6c0d1a